### PR TITLE
Include source behaviour from polls

### PR DESCRIPTION
**Pull request by pzdkn**
# Description

Include the desired behaviour for passing a source argument when using `compose()`:

- We can define constructor for custom Composable, but must leave out source argument
- We can leave out constructor, and use the base one. 

Fixes # issue

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring including code style reformatting 
- [ ] Other (please describe):

# Checklist:

- [ ] I have read the [contributing guideline doc](https://squirrel-core.readthedocs.io/en/latest/usage/code_of_conduct.html) (external contributors only)
- [x] Lint and unit tests pass locally with my changes
- [x] I have kept the PR small so that it can be easily reviewed  
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] All dependency changes have been reflected in the pip requirement files.

### CODE REVIEW

File: squirrel/iterstream/base.py [hunk 1]
```diff
@@ -40,14 +41,20 @@ def __iter__(self) -> t.Iterator:
     def compose(self, constructor: t.Type[Composable], *args, **kw) -> Composable:
         """Compose the items in the stream with items generated by `constructor`"""
         assert "source" not in kw
-        kw["source"] = self
-        return constructor(*args, **kw)
+        if "source" in inspect.signature(constructor).parameters:
+            assert inspect.signature(constructor).parameters["source"].default is None
```
Comment (AlirezaSohofi) [inline squirrel/iterstream/base.py:45]: we should disallow having `source` in the parameters, so raise at `45`.

File: test/test_iterstream/test_stream.py [hunk 2]
```diff
@@ -59,16 +59,39 @@ def test_map(samples: t.List[t.Dict]) -> None:
...
 
     class Add1(Composable):
+        # use default constructor, supported by compose
         def __iter__(self):
             for i in self.source:
                 yield i + 1
 
+    class Add2(Composable):
+        # use constructor without source, supported by compose
+        def __init__(self):
```
Comment (AlirezaSohofi) [inline test/test_iterstream/test_stream.py:72]: please add a case that has arguments in the init

File: test/test_iterstream/test_stream.py [hunk 3]
```diff
@@ -59,16 +59,39 @@ def test_map(samples: t.List[t.Dict]) -> None:
...
+            super().__init__()
+
+        def __iter__(self):
+            for i in self.source:
+                yield i + 2
+
+    class Add3(Composable):
+        # use constructor with source, not supported by compose
+        def __init__(self, source: t.Union[t.Iterable, t.Callable]):
```
Comment (AlirezaSohofi) [inline test/test_iterstream/test_stream.py:81]: please add a case `source=None` in the init. It should fail.

File: squirrel/iterstream/base.py (hunk 1 above)
Comment (AlirezaSohofi) [inline squirrel/iterstream/base.py:45]: Please raise an informative exception

**Review by AlirezaSohofi — CHANGES_REQUESTED** [review]

File: squirrel/iterstream/base.py (hunk 1 above)
Comment (pzdkn) [reply squirrel/iterstream/base.py:45]: If we don't implement a constructor, then the base constructor is used and this one will have a default source set to None. 

As far as I understand we wanted to still support the case of omitting the constructor, wdyt @AlirezaSohofi ?

File: test/test_iterstream/test_stream.py (hunk 3 above)
Comment (pzdkn) [reply test/test_iterstream/test_stream.py:81]: passing `None` should be fine right?

File: squirrel/iterstream/base.py (hunk 1 above)
Comment (pzdkn) [reply squirrel/iterstream/base.py]: Not sure if ValueError is the right Exception here

File: test/test_iterstream/test_stream.py [hunk 4]
```diff
@@ -59,16 +59,62 @@ def test_map(samples: t.List[t.Dict]) -> None:
...
+            super().__init__(source)
+
+        def __iter__(self):
+            for i in self.source:
+                yield i + 3
+
+    class Add4(Composable):
```
Comment (AlirezaSohofi) [inline test/test_iterstream/test_stream.py:88]: this should fail

File: squirrel/iterstream/base.py [hunk 5]
```diff
@@ -41,7 +41,7 @@ def __iter__(self) -> t.Iterator:
     def compose(self, constructor: t.Type[Composable], *args, **kw) -> Composable:
         """Compose the items in the stream with items generated by `constructor`"""
         assert "source" not in kw
-        if "source" in (param := inspect.signature(constructor).parameters) and param["source"].default is not None:
+        if "__init__" in constructor.__dict__ and "source" in inspect.signature(constructor).parameters:
             raise ValueError(f"source argument in {constructor} should be omitted")
```
Comment (AlirezaSohofi) [inline squirrel/iterstream/base.py:45]:
```suggestion
            raise ValueError(f"If the class provides as constructor argument has an __init__ method, it should not have the source argument.")
```

**Review by AlirezaSohofi — CHANGES_REQUESTED** [review]
LG, please also update the documentation.

File: squirrel/iterstream/base.py [hunk 6]
```diff
@@ -38,16 +39,28 @@ def __iter__(self) -> t.Iterator:
         pass
 
     def compose(self, constructor: t.Type[Composable], *args, **kw) -> Composable:
-        """Compose the items in the stream with items generated by `constructor`"""
+        """
+        Compose the items in the stream with items generated by `constructor.
```
Comment (AlirezaSohofi) [inline squirrel/iterstream/base.py:43]:
```suggestion
        Apply the transformation expressed in the `__iter__` method of the `constructor` to items in the stream.
```

**Pull request closed without merging**